### ISSUE TITLE
Beginner questions:  displaying tiff not successfull with imageio-jpeg; metadata in JPEG not read.

### ISSUE THREAD

**Issue by hvdwolf** [haraldk/TwelveMonkeys#555]
Sorry: I could not find a forum, mailing list or something alike.
As I do think these are beginner questions and not bugs, I put 2 in one post thereby only "abusing" one post instead of two.

I just started using the TwelveMonkeys imageIO library, using the 3.6 version, and have 2 issues.

**Not being able to display tiff when using imageio-jpeg.**
In my project in my build.gradle.kts I have the following:
```
    // Twelve monkeys imageio (supports bmp, jpeg, jpeg-2000, PNM (PBM/PGM/PPM/PAM), PSD, TIFF, HDR, IFF, PCX
    // PICT, SGI, TGA, ICNS, ICO & CUR, SVG, WMF)
    implementation("com.twelvemonkeys.imageio:imageio-core:3.6")
    implementation("com.twelvemonkeys.imageio:imageio-metadata:3.6")
    //
    implementation("com.twelvemonkeys.imageio:imageio-jpeg:3.6")
    //
    implementation("com.twelvemonkeys.imageio:imageio-tiff:3.6")
    //implementation("com.twelvemonkeys.imageio:imageio-psd:3.6")
    implementation("com.twelvemonkeys.imageio:imageio-pnm:3.6")
    implementation("com.twelvemonkeys.imageio:imageio-pcx:3.6")
    implementation("com.twelvemonkeys.imageio:imageio-sgi:3.6")
    implementation("com.twelvemonkeys.imageio:imageio-tga:3.6")
    implementation("com.twelvemonkeys.imageio:imageio-icns:3.6")
    implementation("com.twelvemonkeys.imageio:imageio-hdr:3.6")
```
My java project reads images and creates previews of them. When using the `imageio-jpeg` no tiff previews are created and when opening them from my app in a (fullscreen) jframe I do get a `java.lang.NullPointerException`. Somehow the `new ImageIcon`, with or without resizing, is no longer working for tiffs with `imageio-jpeg` enabled.
When disabling the `imageio-jpeg` in my build.gradle.kts, I do see tiff previews and I am able to display tiff images full screen.
Is this a bug or some beginner mistake?

**Reading metadata in jpegs does not work**
I thought that EXIFReader is deprecated and that TIFFReader should be used for all images when reading metadata.

So I created a new test method in my project and did
```
try {
    Directory metadata = new TIFFReader().read(new FileImageInputStream(file));
    for (Entry entry: metadata) {
        //logger.info("identifier {} fieldname {} value {} valueString {}, typename {}", entry.getIdentifier(), entry.getFieldName(), entry.getValue(), entry.getValueAsString(), entry.getTypeName());
        logger.info("fieldname {} value {}", entry.getFieldName(), entry.getValue());
    }
} catch (IOException e) {
    e.printStackTrace();
}
```
When I used this on tiff images it works great. 
When using on jpg's I get:
`javax.imageio.IIOException: Invalid TIFF byte order mark '��', expected: 'II' or 'MM'`

This is happening on all jpegs from multiple cameras and sources.
I fully understand that it is my fault, but I could not find how to get my metadata from my jpegs?

**Comment by haraldk** [comment]
Hi,

I don't have a mailing list or forum, but you might try to ask at StackOverflow, there's a TwelveMonkeys tag you can use which I monitor. 😉

I don't see any reason why the JPEG plugin should interfere with the TIFF plugin. There's no inter-dependencies there, and no related bugs that I know of. Your build.gradle looks good to me, though I'm more familiar with Maven. If you can create a standalone main method or JUnit test that reproduces the problem, I'll look into it. 

For the `TIFFReader`, it can really only read TIFF. It works well with TIFF files, because... well, they are TIFF. 😀 Exif data is also TIFF. But, it does not understand the JFIF wrapper stream in your JPEG files. You need to first locate the Exif data inside the JFIF stream, and pass the stream directly at that position (or a wrapped byte array stream or similar). 

Version 3.6 of the JPEGImageReader does expose some data from the Exif metadata in the ImageIO metadata standard format, which may be easier to access/more standard. But if you want to parse the JPEGs yourself to get full access to all Exif information, you could look at the [source code for the JPEGImageReader](https://github.com/haraldk/TwelveMonkeys/blob/ed14b971991951379365f975bd6d9caaea8a836b/imageio/imageio-jpeg/src/main/java/com/twelvemonkeys/imageio/plugins/jpeg/JPEGImageReader.java#L901) itself, to see how to read/parse the Exif data.


Best regards,

-- 
Harald K

**Comment by hvdwolf** [comment]
> For the `TIFFReader`, it can really only read TIFF. It works well with TIFF files, because... well, they are TIFF. 😀 Exif data is also TIFF. But, it does not understand the JFIF wrapper stream in your JPEG files. You need to first locate the Exif data inside the JFIF stream, and pass the stream directly at that position (or a wrapped byte array stream or similar).
> 
I can really bang my head to the wall :angry: . After more than 15 years working with metadata (but only recently with java): Mixing up Exif with jpeg: A blind spot for several hours.

For the rest: I will try that tonight/tomorrow.

**Comment by hvdwolf** [comment]
Feedback on **Not being able to display tiff when using imageio-jpeg**

Below 2 lines from my build.gradle.kts
```
    implementation("com.twelvemonkeys.imageio:imageio-tiff:3.6")
    implementation("com.twelvemonkeys.imageio:imageio-jpeg:3.6")
```
TIFF only works when I put the jpeg dependency after the tiff dependency in the gradle file. I really have no idea why.
I work on 64bit Ubuntu Linux 18.01 using OpenJDK V8 to stay compliant with quite some Windows/MacOS users who did not make the step to 9-14 yet, and I did find some remarks in the issues and stackoverflow about openjdk "less compatible" than Oracle's jdk due to several libjpeg dependencies, so I compiled also with the Oracle jdk, but there is no difference. The jpeg dependency has to come after the tiff dependency.
Compiling with OpenJDK 11 makes no difference either. The jpeg dependency always has to come after the tiff dependency.

Being a newbie  in java I do not even know how to make a JUnit test, but coming weekend I have more time. So I will spend some more time on my own education.

**Comment by haraldk** [comment]
Hmmm.. Strange.

I'm just using Maven and usually declare the dependencies alphabetically (JPEG before TIFF). No problem with that. Also, these should be just runtime (or perhaps test time) dependencies, so I don't see why the order matters. And the runtime order may even be different than in the build file. Maybe you need to ask in a Gradle forum for more details on that.

(The only dependency between the modules is that the `imageio-tiff` module has a `test` scope dependency to the `imageio-jpeg` module, to allow testing some TIFF files with JPEG compression. But this should not be relevant for your Gradle build, using the pre-built artefacts. It is only relevant if you try to build the TwelveMonkeys ImageIO library from source, and if anything, it should dictate JPEG before TIFF...)


PS: Are you by any chance creating a "fat" JAR or other kind of bundle where you re-pack the contents of the JARs?

-- 
Harald K

**Comment by hvdwolf** [comment]
I do compile the imageio sources in my project and do not use the ready made jars. I will try those as well this weekend.
And yes: I also have the dependencies alphabetically. That's why jpeg came before tiff :wink:
But after running into my strange problems I moved tiff directly behind jpeg, removed jpeg and finally placed jpeg behind tiff.

I do create a fat jar using the shadow plugin
```
    //Shadow plugin for fatJar
    id("com.github.johnrengelman.shadow") version ("5.0.0")
```
but I also build distributions (actually gradle does that automatically): both as fat jars and as {/bin, /lib} with my jar in the /bin and all the dependency jars in the /lib folder, including the imageio jars (but again: currently compiled from source)

**Comment by haraldk** [comment]
So... You have basically created a new build script that I haven't seen, using a build tool I don't know, and are asking me why it does not work...? 😉

I suggest you use the pre-built binaries, they are available through Maven Central and should be easy to get using most build tools. If you have further build issues, consult a Gradle forum.

For creating fat JARs or other re-packing of the ImageIO plugins, it is important that you (or the fat JAR plugin you use) understand the [JAR Service Provider Interface (Spi)](https://docs.oracle.com/javase/tutorial/sound/SPI-intro.html) mechanism. See #502 or [this StackOverflow answer](https://stackoverflow.com/a/32652548/1428606). You can't just copy everything into one folder and hope it works (well, you can always hope, but it *won't work*). 😀

TL;DR: You need to merge the entries in the `META-INF/services` folder inside the JARs. Consult the Shadow plugin documentation for details.
 

Best regards,

-- 
Harald K

**Comment by hvdwolf** [comment]
Quick lunch tryout.
I now replaced the "to be compiled imageio" sources with all the 3.6 jars in a libs folder in my project.
in build.gradle.kts Kotlin style, that is:
`implementation(fileTree(mapOf("dir" to "libs", "include" to listOf("*.jar"))))`

In my code I simply use: `BufferedImage img = ImageIO.read(new File(file.getPath().replace("\\", "/")));`
But now again, I get the error `java.lang.NullPointerException` and only when reading tiffs and on windows10 and Linux (could not test MacOS right now, but have no different expectation)

i also tried the approach of:
```
ImageInputStream iis = ImageIO.createImageInputStream(is);
ImageReader r = getImageReaders(iis).next();
..etcetera .. etcetera..
```
And that again does work for png and jpg, but not for tif.

**Comment by haraldk** [comment]
As mentioned, I cannot help you debug your Gradle build. You need to consult a Gradle forum for that.

Your code should work (given the file exists and is a valid TIFF file, and the TIFF plugin is properly installed). `ImageIO.read(...)` will return `null` for a format it does not recognise, so you need to handle that. I believe this is the source of your `java.lang.NullPointerException`. I have not seen the stack trace (and I don't need to, unless it includes frames with `com.twelvemonkeys.*`). 

Try printing the output of `ImageIO.getReaderFormatNames()` to see the formats ImageIO can read at runtime. Show the results. Does it include "TIFF"? Check your class path, to see if the `imageio-tiff.jar` is really available at runtime.

Yes, PNG and JPEG will work out of the box, because those formats are bundled with your JDK/JRE. You can verify if the TwelveMonkeys JPEG plugin is installed, by using the code sample from the [Installing](https://github.com/haraldk/TwelveMonkeys#installing) section in the readme (same will work for TIFF, by replacing the format name).


-- 
Harald K

**Comment by hvdwolf** [comment]
Thanks. I understand that it is not possible to troubleshoot without knowing the details.
I do use a combination of IntelliJ with gradle. It would require a lot of screenprints.
And it started "simple" with the jpeg/tiff question and if there was some inter-dependency.
If I put jpeg before tiff, I do not have tiff

check:
```
        String[] readerFormatNames = ImageIO.getReaderFormatNames();
        for (String reader : readerFormatNames) {
            logger.info("reader {}", reader);
        }

```

tiff before jpeg | tiff after jpeg
-----------------|---------------
reader JPG  | reader JPG
reader jpg | reader jpeg
reader tiff | --
reader bigtiff | --
reader bmp | reader bmp
reader BMP | reader BMP
reader gif | reader gif
reader GIF | reader GIF
reader WBMP | reader WBMP
reader png | reader png
reader PNG | reader PNG
reader BIGTIFF
reader JPEG | reader JPEG
reader BigTIFF | --
reader tif | --
reader TIF | --
reader TIFF | --
reader jpeg | reader jpeg
reader wbmp | reader wbmp
--  | reader JPEG-LOSSLESS
 -- | reader jpeg-lossless

The strange thing is that I have jpeg-lossless when I do not have tiff, which might be some wrong interpretation of tiff-lzw and/or packbits.
And somehow the iff/pcx/pict/pnm/sgi/tga are not included either, not in any combi
It looks like the default java8 formats with tiff included if I put tiff after jpeg

Tested with OpenJDK 1.8_265, Oracle jdk 1.8_265,  and column 2 with precompiled jars included for both openjdk and oracle..

I will close the ticket now and try to analyse what is going wrong.

**Comment by haraldk** [comment]
The fact that you don't have all the iff/pcx/pict/pnm/sgi/tga formats available and only either the TIFF or the JPEG plugin (only the TwelveMonkeys JPEG plugin supports jpeg-lossless), suggests again that the problem is the bundling as mentioned [in this comment](https://github.com/haraldk/TwelveMonkeys/issues/555#issuecomment-670477425). The fat JAR packager does not correctly merge entries in `META-INF/services`, meaning you are left with only a single plugin. 

-- 
Harald K

**Comment by Schmidor** [comment]
We had those services issues back when we did create fat jars for a webstart. Our merging code for gradle might be helpfull :)

```
task fatJarWebstart {
  doLast {
    def serviceDir = file("$buildDir/tmp/bigJar/META-INF/services")
    serviceDir.deleteDir()
    serviceDir.mkdirs()
    // copy all service files from deps to buildDir
    for(file in configurations.runtimeClasspath) {
      zipTree(file).matching{ include 'META-INF/services/*' }.each { f ->
        new File(serviceDir, f.name) << f.getText("UTF-8") + '\n'
      }
    }

    ant.jar (destfile: 'build/webstart/lib/libs.jar', duplicate: 'preserve'){
      String cp = configurations.runtimeClasspath.asPath.toString()
      cp.tokenize( ';' ).each {
        def jar = file(it)
        if (!jar.name.startsWith( 'bc' )){
          zipfileset (src: jar.path){
            exclude (name: 'META-INF/*.SF')
            exclude (name: 'META-INF/*.DSA')
            exclude (name: 'META-INF/*.RSA')
            exclude (name: 'META-INF/README.*')
            exclude (name: 'META-INF/NOTICE.*')
            exclude (name: 'META-INF/NOTICE')
            exclude (name: 'META-INF/LICENSE.*')
            exclude (name: 'META-INF/LICENSE')
            exclude (name: 'META-INF/MANIFEST.MF')
            exclude (name: 'META-INF/maven/**')
          }
        }
      }
      delegate.manifest {
        attribute(name: 'Permissions', value: 'all-permissions')
        attribute(name: 'Codebase', value: '*')
        attribute(name: 'Implementation-Vendor', value: 'Sun Microsystems, Inc.')
        attribute(name: 'Implementation-Title', value: 'com.sun.media.imageio')
        attribute(name: 'Implementation-Version', value: '1.1')
        attribute(name: 'Specification-Vendor', value: 'Sun Microsystems, Inc.')
        attribute(name: 'Specification-Title', value: 'Java Advanced Imaging Image I/O Tools')
        attribute(name: 'Specification-Version', value: '1.1')
        attribute(name: 'Extension-Name', value: 'com.sun.media.imageio')
      }
    }
    ant.jar(update: "true", destfile: 'build/webstart/lib/libs.jar') {
      fileset(dir: "$buildDir/tmp/bigJar", includes: 'META-INF/services/*')
    } 
  }
}
```
There were some specifics for jai-imageio, which needs some manifest entries, and not including bouncy castle in the merging to preserve the signed security provider classes.

**Comment by hvdwolf** [comment]
Thanks both for your time and patience.
I now fully understand what the problem is. Indeed the distribution version with all jars separately does work correct.

I do understand the script from Schmidor and what it is doing. The point is that the `build.gradle` syntax is different from the kotlin `build.gradle.kts` syntax, and I did not find the correct documentation yet, describing how to do the same. But that is my problem.
I now solved it in a likewise "external" shell scripted way by extracting all lines from all created readerSpi and writerSpi from the normal build separate jars, write these lines to a single readerSpi and writerSpi and simply add these 2 Spi files to the already compiled fatJar.

Definitely not how it should be done, but it works. Now I will search how to do above script nicely in the build.gradle.kts script.
